continuous_integration/scripts/test_report.py

```diff
@@ -258,40 +309,35 @@ def download_and_parse_artifacts(
     ndownloaded = 0
     print(f"Downloading and parsing {nartifacts} artifacts...")
 
-    cache: shelve.Shelf[pandas.DataFrame | None]
     # FIXME https://github.com/python/typeshed/pull/8190
-    with shelve.open("test_report") as cache:  # type: ignore[assignment]
-        for w in workflows:
-            w["dfs"] = []
-            for a in w["artifacts"]:
-                url = a["archive_download_url"]
-                df: pandas.DataFrame | None
-                try:
-                    df = cache[url]
-                except KeyError:
-                    xml = download_and_parse_artifact(url)
-                    if xml:
-                        df = dataframe_from_jxml(xml)
-                        # Note: we assign a column with the workflow timestamp rather
-                        # than the artifact timestamp so that artifacts triggered under
-                        # the same workflow can be aligned according to the same trigger
-                        # time.
-                        df = df.assign(
-                            name=a["name"],
-                            suite=suite_from_name(a["name"]),
-                            date=w["created_at"],
-                            url=w["html_url"],
-                        )
-                    else:
-                        df = None
-                    cache[url] = df
-
-                if df is not None:
-                    yield df
-
-                ndownloaded += 1
-                if ndownloaded and not ndownloaded % 20:
-                    print(f"{ndownloaded}... ", end="")
+    for w in workflows:
+        jobs_df = get_jobs(w)
+        w["dfs"] = []
+        for a in w["artifacts"]:
+            url = a["archive_download_url"]
+            df: pandas.DataFrame | None
+            xml = download_and_parse_artifact(url)
+            if xml is None:
+                continue
+            df = dataframe_from_jxml(cast(Iterable, xml))
+            # Note: we assign a column with the workflow timestamp rather
+            # than the artifact timestamp so that artifacts triggered under
+            # the same workflow can be aligned according to the same trigger
+            # time.
+            job = jobs_df[jobs_df["suite_name"] == a["name"]]
+            df2 = df.assign(
+                name=a["name"],
+                suite=suite_from_name(a["name"]),
+                date=w["created_at"],
+                html_url=job["html_url"].unique()[0],
```

Comment:
Given the fragility of the construction of `"suite_name"`, I'd recommend making this robust to the case where `job` has no entries.

Reply:
I'll raise an assert if this is empty. Shouldn't happen.

Reply:
It's the type of thing that can happen if the test matrix changes (which has happened in the past)